### DISCUSSION TITLE
Feedback on primary constructors

### DISCUSSION THREAD

**Discussion by KathleenDollard** [dotnet/csharplang#7667] [General]
C# 12 introduces primary constructors for non-record types: https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#primary-constructors

What else would you like to see and what are the important scenarios to you? A gesture that creates properties in non-record types? Support for immutability in primary constructor parameters? Something else?

**Comment by DavidArno** [comment]
> .. Support for immutability in primary constructor parameters? ..

I'm really puzzled as to why this question is being asked. It should have been super clear to MS from feedback a month or two back that lots of developers were deeply unhappy with the way mutable "captures" were the default behaviour for non-record PCs and how any other behaviour had to be achieved via extra code and custom analyzers. 

We face the situation in my present company whereby the consensus in the development department is that this feature is so far removed from being "fit for purpose", we will be implementing a custom analyzer to block the use of this feature until PCs are "properly" implemented. To put it into context, we already have stories in our backlog to update our code base to C#12 and a "brown bag" session planned to show off the other new features later this month. These are developers that are normally enthusiastic about new releases. This is the first time for any C# feature, in any release, that the team has even remotely contemplated blocking it. 

Just get on with supporting a means of those PC parameters being "captured" via private readonly fields with the minimum code ceremony possible, as quickly as possible, please.

**Reply by qrli** [reply]
While I'm not emotional on this, since this is a feedback thread :-), I'm with DavidArno. I think mutable primary constructor parameter is a bad choice.

My feeling is that, the designers eye more on the small classes scenario, where primary constructor fits nicely just like records, or senior library developers who absolutely know what they are doing. However, my gut feeling is that majority developers will use it for other scenarios: Half will be for dependency injection, 1/4 for domain objects with single constructor (DTOs use case can be replaced by records). Both are often big classes -  there are more average developers than senior ones. And some classes will even span over multiple files by partial. And the code maintainer rotates. Mutation from a far away location is as tricky as mutation made by another thread, especially worse when you are reading an unfamiliar codebase.

**Comment by zms9110750** [comment]
i want this:
```csharp
class State()
{
	private readonly TreeNode<State> tree=new TreeNode<State>(this); 
	public State Substate => tree.Parent.Value;
}
```
You see, the main constructor with parameters. The assignment of fields using parameters is already in the execution phase of the constructor.
Since `this` can be used during the execution phase. So why can't classes that use the main constructor use `this`

**Reply by CyrusNajmabadi** [reply]
Because fields are assigned prior to the method body being executed.  Primary constructors don't change that.

**Reply by ufcpp** [reply]
There is a suggestion in #1026, but is it not being considered very positively?

**Reply by HaloFour** [reply]
That proposal hasn't really gone anywhere.  I think the issue is that doing so will silently change the order of when certain fields are initialized, which will likely be unintuitive and surprising.  It also creates a dependency problem as any member initializers that depend on that late-initialized member will also have to be late-initialized, including in derived types and possibly across assembly boundaries.

**Comment by egvijayanand** [comment]
How to invoke a method like `InitializeComponent()` which takes no parameters and returns nothing (this does initialize some UI components) in the context of Primary Constructor?

In general, the question here is how to do something more with a PC.

Similar to partial methods, define a construct that gets hooked up in the PC body during auto code gen.

**Reply by HaloFour** [reply]
I think there's value in supporting a primary constructor body.  Java accomplishes this with records using the following syntax:

```cs
record Point(int x, int y) {
    Point {
        // code goes here
    }
}
```

**Reply by 333fred** [reply]
I think one of the points we're unsure of here is: primary constructor bodies, or final initializers? Both are valid things to want, but do we want to add both?

**Reply by HaloFour** [reply]
Do we have proposals open for final initializers?

I assume that means a method declared on the type that the compiler calls after all properties are initialized?

**Reply by 333fred** [reply]
I don't think we have anything written up, but yes, that's the general idea.

**Reply by KennethHoff** [reply]
~~By "final initializer" do you mean something that happens after the object initializer ( which itself happens after the constructor)? If so that would be very nice in many scenarios.~~

Already answered while I was writing it..

**Reply by qrli** [reply]
> I think one of the points we're unsure of here is: primary constructor bodies, or final initializers? Both are valid things to want, but do we want to add both?

I'd think `init` properties are like another form of constructor parameters, then final initializer is the only thing necessary.

**Reply by HaloFour** [reply]
> I think one of the points we're unsure of here is: primary constructor bodies, or final initializers? Both are valid things to want, but do we want to add both?

I'm going to vote that we do need both.

We need final initializers, because they are the only way to validate that init members have been populated properly.  However, they affect the public API of the type.  Adding or removing a final initializer is a source and binary breaking change, affecting all potential consumers.

That's why we also need constructor bodies.  They are an implementation detail and can be added/removed/amended at will without impacting the public contract.

**Reply by qrli** [reply]
>  Adding or removing a final initializer is a source and binary breaking change, affecting all potential consumers.

That's a very good point.
 
However, I don't think average developers will put enough caution into this difference. So I don't think the language should leave it as breaking change, regardless doing both or not, otherwise it will be a source of bugs and dev teams will simply ban the final initializer feature after being bitten. I'd think the compiler and runtime can do things to avoid such breaking behavior, while not sacrificing performance. There are already other features and use cases, e.g. inserting instrument points.

**Reply by HaloFour** [reply]
@qrli 

> So I don't think the language should leave it as breaking change, regardless doing both or not, otherwise it will be a source of bugs and dev teams will simply ban the final initializer feature after being bitten.

I'd be very curious of an implementation that could accomplish that.  I would expect that a final initializer would be a method declared on the type that the consumer must call post-construction/initialization.  A type without a final initializer would lack that method.

**Reply by qrli** [reply]
@HaloFour I'm not a compiler guy. But I know there are features e.g. to insert a call somewhere. The call could be optimized away by JIT when possible. Or, the call points at some code which calls final initializer when it exists, while does nothing when it does not exist.

**Reply by HaloFour** [reply]
The compiler isn't the problem, the runtime would be.  The JIT can optimize a call away or elide it if it does nothing, but the runtime doesn't have the facility to handle a method call to a method that doesn't exist.  Could it?  Maybe.  That feels like a really awkward feature to me.

**Reply by qrli** [reply]
> to handle a method call to a method that doesn't exist. 

That's why I said using an intermediate call, say `InvokeFinalInitializer(object obj)` - which is another function, whose implementation may be generated by runtime, while compiler merely generates that function with a dummy implementation - e.g. empty. Or better, the intermediate function could be a built-in function of the runtime.

Without special runtime support, the intermediate function would have to resort to reflection.

**Reply by HaloFour** [reply]
While I'm sure that's technically possible, I also don't think it would be palatable.  There aren't any other features like that in .NET.  It also doesn't solve for the addition of of a final initializer, unless the compiler emits such a call for every instance that is initialized.

**Reply by erik-kallen** [reply]
In Scala, you can just write whatever statement you want directly in the class body.

**Reply by egvijayanand** [reply]
Unless there's an option to extend this PC with a body, the applicability is limited to the Model and Service types and definitely not to UI modeling such as XAML and RAD interfaces.

**Reply by egvijayanand** [reply]
I'm proposing an approach similar to that of [Module Initializer](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.moduleinitializerattribute?view=net-8.0).

Define a Type Initializer attribute and allow the programmer to mark any (private instance) method with this attribute (a method that takes no parameters and returns void) since the whole type has access to the parameters of the PC, there's no need to take any explicit parameters.

The automation needs to invoke the method with this attribute in the source-generated constructor. And validate all the rules that apply to the constructor method as if it is the Proxy Constructor (even this abbreviates to the same PC).

@MadsTorgersen @KathleenDollard

**Comment by KennethHoff** [comment]
Personally I think we should stay away from adding features to the PC (**P**rimary **C**onstructor) parameter list that makes little-to-no sense in any other parameter list.

`readonly` parameters? Does that make sense to add to any ol' parameter list? I think so, therefore `readonly` makes perfect sense for PCs - even if `readonly` support for non-PC parameter lists don't arrive in the same release/ever.

Creating a public property? Does that make sense to add to any ol' parameter list? I think no. Feels out of scope. That said though; Could the syntax for props-in-PCs translate cleanly to some other tangentially related concept in method signatures? Possibly, but I can't think of any right now.

I think that we should not confuse Primary Constructors with "Simplifying data-modelling" because they are tangentially related at best.

**Reply by AndersKehlet** [reply]
Making PC parameters public can also easily be done with a generator, so I'd rather the team work on other things.

**Reply by DavidArno** [reply]
@AndersKehlet , "_if all you have is a hammer, everything looks like a nail_"

Just about any feature you can think of can be done with generators. Unions? Why bother, just use a generator:
```cs
[Union]
partial class Option<T>
{
    [UnionCase]
    private None __none;

    [UnionCase]
    private T __some;
}
```
And the reason why we don't just use a generator is that the code is pig-ugly, complicated and such non-standard, non-idiomatic code leads to splintering into dev camps. Just because a generator can be used, doesn't mean it should.

Generators should be seen as meeting the requirements of domain-specific features, not to bypass proper language design for generic features.

**Reply by HaloFour** [reply]
> `readonly` parameters? Does that make sense to add to any ol' parameter list?

I think so, yes.  Given the prevalence of captures in C# I think the ability to mark specific parameters and locals as `readonly` would be beneficial to avoid accidental mutation of state.  I'd prefer this to apply across the entire language, but given the scope of primary constructor parameter captures I think it makes sense to start there.

> Creating a public property? Does that make sense to add to any ol' parameter list?

It already makes sense with primary constructions, given records.

The question (IMO) is whether language features are absolutely necessary for this, or if it could be solved in a satisfactory manner via analyzers and source generators.  Like records, I do think this is deserving of first class support.

**Reply by AndersKehlet** [reply]
@DavidArno 

Generating public members for PC parameters is _trivial_ and it doesn't make the user code ugly.

Generating unions is technically possible for reference types, though I think the shown approach is questionable.

Generating unions that are also checked for exhaustiveness is, again, technically possible though in addition to the generator it requires a decidedly non-trivial analyzer.

Generating struct unions is... maybe possible? You could have an interface as the common type, but then you can't really guarantee the hierarchy is closed, so you lose exhaustiveness. There'd also be boxing.

The language proposal unifies all of this with a consistent syntax that is considerably simpler than what users can do.

**Reply by CyrusNajmabadi** [reply]
@AndersKehlet I agree. :)

**Reply by AndersKehlet** [reply]
@CyrusNajmabadi - So... unions when? 😄

@DavidArno - btw, the reason I think your union example is bad, is that I wrote a union generator two weeks ago and the user code looks like this:

```csharp
[AutoClosed]
public partial record Option<TValue>
{
    partial record Some(TValue Value);

    partial record None;
}
```

Needless to say, making the declaration syntax "pretty" is not why I want it as a language feature. Your other concerns are slightly more reasonable, though I think you exaggerate the issue with "non-idiomatic" code. I may have preferences, but I'm not gonna fortify myself in a camp. If I see an option or maybe type in a project, I'll immediately have a pretty good inkling of what it does, and I'll be comfortable using it (even though it'll likely be a mere pseudo-union, an abominable pretender).

**Comment by konrad-jamrozik** [comment]
Asking for three features working together: support for always-enforced construction preconditions, support for `required init` immutability, and support for `with` working with these features.

1. I would like to be able to capture in a primary constructor arbitrary preconditions/invariants that must hold when a given immutable class is constructed.

``` csharp
public class Triangle(int x, int y, int z)
{
  primary_ctor_body()
  {
    // if for any of the x, y, z the sum of the two other values is shorter or equal, 
    // throw an exception the triangle construction precondition / invariant was violated 
    // because it is not possible to create a valid triangle with these lengths.
  }
}
```

2. I would also like to be able to, **in effect** (actual syntax can differ), denote the primary ctor params as `required init` for reasons listed in [Object Initializers with the `init` accessor](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#object-initializers-with-the-init-accessor) (which was added based on my [docs issue #35014](https://github.com/dotnet/docs/issues/35014)), specifically:
  > Required init-only properties support immutable structures while allowing natural syntax for users of the type.

Now, with 1. and 2. together I can construct an **immutable** object (due to `init`), with the flexibility of using object initializers for some of the members, with the ability to always ensure everything is well-constructed (due to `required`) and ensuring the object obeys the construction preconditions (as they are enforced by the primary ctor which is always called). 

3. If then I could also use [`with` expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression) like `Triangle newTriangle = oldTriangle with { y = 10 }` which would automatically call the primary ctor with appropriate arguments and check the preconditions, creating new immutable triangle, that would be awesome!

**Reply by vladd** [reply]
As the property initializers run strictly after the constructor finishes (including derived ones), it seems that you are asking for a post-construction, post-property-initialization method. I think I’ve seen a proposal about such a method, but it was separate from the primary constructors.

**Reply by konrad-jamrozik** [reply]
Thank you @vladd for the clarification. Bottom line, conceptually, I am hoping for an ability to enforce "object construction invariants". It pains me a bit that currently if I want to do it, I cannot use primary constructors, because they don't have the body of a normal boilerplate-heavy constructor where I would usually put such invariants.

It would also be nice to make `with` expressions work with that, but as you just clarified, this doesn't seem to be easy to do because the property initializers run after constructors, and I understand the prop initializers are used by the `with` expressions.

Overall, looks like the language is adding various nifty features but because they don't work well together, I cannot really use them. Bummer.

**Comment by betsson-vyza** [comment]
Now anything can be initialized anywhere. Feels like C# is moving towards javascript.

**Comment by yugabe** [comment]
It would be nice if the behavior of properties could be defined on the primary constructor parameters themselves as closely as possible to the normal auto-properties themselves. I mean any or all of:
- visibility,
- mutability (`readonly`), although I believe this should've been the default anyways (I don't see a use case for the contrary),
- `get`-only or also `set`table (optionally limiting visibility on the `set`ter); `init` makes no sense in this case, I believe.

Probably by reusing the relevant property syntaxes. Maybe:
```csharp
public class Person(
  public string Name { get; private set; }
)
{ }

```

I believe that's as far as I would go. If you would need additional possibilities, like inheritance modeling with virtual members, there's the "classic" way to define properties.

Also, I believe final initializers would be nice to have.

~~My most pressing problem with primary constructors at this point is that there's no control over the capturing, thus, it's relatively easy to reference the constructor parameter instead of a property **by mistake**, for example:~~

```csharp
public class Person(string name)
{
  public string Name { get; set; } = name;
  public override string ToString() => name; // Oooops, this should've been "Name".
}
```


As per @CyrusNajmabadi, this seems like a non-issue.

**Reply by Krzysztof318** [reply]
I don't think putting full property declaration in primary constructor is a good idea.
Mutability of course but others no.

**Reply by CyrusNajmabadi** [reply]
@yugabe That mistake seems hard to make.  First, the IDE won't even recommend 'name' at all, once it is captured.  Second, if you were to write that, you are immediately told:

```
Parameter 'string name' is captured into the state of the enclosing type and its value is also used to initialize a field, property, or event.
```

So i'm not sure how you're ending up in that state.

**Reply by yugabe** [reply]
Thanks @CyrusNajmabadi, I didn't know there were analyzers for these. I remember there was talk about this around design, but wasn't solved yet then. Consider my fears alleviated.

**Reply by CyrusNajmabadi** [reply]
@yugabe You're welcome :)

**Comment by hannah23280** [comment]
I wonder if too late to suggest parameter in the PC to be public (instead of private) by default, similar to that of record type.  Immutability can be enabled explicitly.

**Reply by CyrusNajmabadi** [reply]
The feature has shipped. It is very intentional that it does not create a public member.

**Comment by glen-84** [comment]
As I mentioned [here](https://github.com/dotnet/csharplang/issues/2691#issuecomment-1490951234), I'd like to be able to create a field from the constructor parameter, so that it can be accessed with `this`.

`field` keyword, private and read-only by default.

```cs
public sealed class Example(field IMediator mediator)
```

**Comment by HaloFour** [comment]
I'd like to see the `readonly` modifier used to mark the parameters as readonly when they are captured.

I'd also like to see accessibility modifiers applied that would promote the parameter to a named member within the type.

The two could be combined to promote the parameter to a `readonly` member.

```cs
public class Foo(
    string foo, // regular mutable capture
    readonly string bar, // readonly capture,
    public string baz, // public mutable field
    public readonly string qux // public readonly field
) {

    // ...
}
```

Inline declarations of properties could be interesting also, although I wonder if that's a step too far, especially as I imagine that there's a limit to how complicated a property could be.  An auto-property could be palatable, but a partial auto-property or a full property?

**Reply by DavidArno** [reply]
~Sorry, @HaloFour, but the idea that I'd have to use `private readonly` to achieve a read-only capture is just awful. Private is the default throughout the language: let's keep it that way.~

Edits to HaloFour's comment addresses this point, so it's retracted.

**Reply by HaloFour** [reply]
I was suggesting two separate features there, that could be combined.  `readonly` to achieve readonly captures, and accessibility modifiers like `private` to achieve promotion to a proper field.  If you wanted readonly captures, but not a proper field, you would only use `readonly`.

Updated example to show some combinations.

**Reply by bzbetty** [reply]
```
public Foo(init int bar) {
   public int Bar {get; private set; } = bar;
}
```

wondering if `init` could be used to limit usage of the parameter to initialization only (eg can't be used in another method)\

**Reply by glen-84** [reply]
Maybe `public` and `protected` should create properties, and `private` should create fields?

**Comment by FaustVX** [comment]
For those interested by this feature, but want more control over the generated field, I've created a NuGet package which allows you to put `[Field]` attribute on a primary contructor parameter, to generate a `private readonly` field. and also to forbid you to use the parameter inside the `class`/`struct`.
NuGet: [FaustVX.PrimaryParameter.SG](https://www.nuget.org/packages/FaustVX.PrimaryParameter.SG)
GitHub: [github.com/FaustVX/PrimaryParameter](https://github.com/FaustVX/PrimaryParameter)

**Reply by 333fred** [reply]
@FaustVX interesting generator. However, I'd recommend coming and talking to us on discord, https://discord.gg/csharp, to address some anti-patterns in your generator.

**Reply by DavidArno** [reply]
So now I have to mark every class that uses PC's with `partial` in order to achieve a read-only field? Thanks, but no thanks. ☹️

**Reply by HaloFour** [reply]
Worse than putting `readonly` on every parameter?  Such a source generator could be configurable that `partial` would be the only keyword you'd need to add in order to change every primary constructor argument into a readonly field or capture.

**Reply by DavidArno** [reply]
Having the `partial` keyword cause the parameters to become readonly is a perfect example of violating the principle of least astonishment. `readonly` tells me what it's doing. `partial` just tells me, "_here be dragons: some weird-shit magic might be happening. If you didn't write this code, good luck working it out 😈_"

**Reply by HaloFour** [reply]
That sounds like more of an indictment of source generators.  I'd suggest that you'd probably want to pair that with a custom attribute that describes what the source generator is going to do, even if it's not an opt-in scenario and would be an error to omit the attribute.

**Reply by taori** [reply]
@333fred Is there any chance there is more work planned for primary constuctors? I really enjoy them from a "less boilerplate" pov, but using them with lower camelcase is something i very much dislike. Sometimes i have 5-15 DI services for UI classes, and from an intellisense perspective _someService just gets my intellisense into a direction way faster. For that sole reason i often find myself reverting primary constructors to normal ones (which is sad, because i really would love to drop the boilerplate)

**Reply by CyrusNajmabadi** [reply]
@taori There is more we're going to look into for primary constructors.

**Reply by CyrusNajmabadi** [reply]
> and from an intellisense perspective _someService just gets my intellisense into a direction way faster.

See discussion below about how a simple source generator would be able to take care of this for you.

**Reply by taori** [reply]
@CyrusNajmabadi Thanks. I've seen it, but i wanted to give the feedback, since adding dependencies is not always a way that is available in every development environment

**Reply by CyrusNajmabadi** [reply]
Not sure what you mean by adding a dependency :)   SGs are something you can just author.  They only generate code you would write yourself.  Just in a way that is automated so you don't have to manually do it :)

**Reply by taori** [reply]
Oh i just meant adding a nuget package which has the source generator as "dependency". obviously it's not really a dependency per se, but a package that is added to the project. I sometimes have customers who are very restrictive with package references

**Reply by CyrusNajmabadi** [reply]
@taori this would be for your own benefit.  This would not be shipped to your customers.  This is about addressing hte point you made about:   Sometimes i have 5-15 DI services for UI classes, and from an intellisense perspective _someService just gets my intellisense into a direction way faster.

For your local library/app development you'd use a generator to take care of that drudgery.  But you wouldn't be shipping anything new (esp. nuget-wise) to your customers.

**Reply by CyrusNajmabadi** [reply]
Note: This is how Roslyn itself works.  We have generators that take care of stuff for us.  But we dont' ship the generators to customers.  We only ship our normal compiled libraries, which contain the compiled *outputs* of the generator.

**Comment by wiz0u** [comment]
Can we have a #pragma or something, such that the fields names can get a prefix underscore?

for example, `public class Person(string name)` would declare the field as `_name` instead of `name` from inside the class.
(but from outside the class, the constructor parameter should still appear as `name` in Intellisense and such)

Alternatively, we could declare `public class Person(string _name)` and have a way to have the initial underscore be removed from the Intellisense constructor parameter names (cleaner).
Could also be something like `public class Person([DisplayName("name")] string _name)` to have more control on the "outside" name of that parameter.

**Reply by HaloFour** [reply]
You could use a source generator that automatically emits code to initialize fields using any name you want.  It could be paired with an analyzer that then makes it an error to reference the parameter as a capture outside of the construction lifecycle.

**Reply by wiz0u** [reply]
I see C# 12 primary constructors as syntactic sugar to have trivially simple automatically initialized fields.
Having to write a source generator to address something many developers use (starting their private fields with an underscore) sounds no longer trivial.

**Reply by CyrusNajmabadi** [reply]
> I see C# 12 primary constructors as syntactic sugar to have trivially simple automatically initialized fields.

That is not correct.  They are just another way of defining a constructor with parameters.  Just with parameters that can be used by the members nested within the type/cosntructor.

**Reply by CyrusNajmabadi** [reply]
> Having to write a source generator to address something many developers use (starting their private fields with an underscore) sounds no longer trivial.

It's trivial to just use that (and have other developers use it) once written once.

**Comment by erik-kallen** [comment]
Just want to add some positive comments among all complaints. I like the feature. It reduces the ceremony for DI, and the lack of readonly isn't too big of a deal. Just don't mutate the value, and you'll be fine. I don't know if the `readonly` modifier in any DI'd field anywhere in my code ever has saved me from accidentally mutating it. It's just something you don't do because there is no good reason to do it.

**Reply by DavidArno** [reply]
Ah, the old "you don't need a seat belt.  Just don't crash your car, and you'll be fine" line of reasoning 😂

**Reply by CyrusNajmabadi** [reply]
Given such a small incidence of this actually happening, I'd say that is fairly accurate. You don't need every safety measure. Not if it's not really that necessary or valuable.

**Reply by erik-kallen** [reply]
@DavidArno I do use a seat belt, but I do not use a helmet when driving. I assume you do.

**Reply by ds1709** [reply]
Besides of mutability `readonly` modifier is also declarative. You can see clearely what fields a injected dependencies, and what are the mutable state of class. Also, it's good if you are a single developer and you know what field you can change, and what you don't. But working in team, someone can write code, that changes field value.

**Reply by ds1709** [reply]
@CyrusNajmabadi 
>Given such a small incidence of this actually happening, I'd say that is fairly accurate. You don't need every safety measure. Not if it's not really that necessary or valuable.

Why `readonly` keyword exists then? What was the point of introducing `in` modifier for method parameters?
from MSDN
> `in` means the method reads, but can't write the value of the argument

Does it have any sence if you can just not change parameter value? Or is it something else?

It all seems just like inconsistency of language evolve strategy.

**Reply by CyrusNajmabadi** [reply]
> What was the point of introducing in modifier for method parameters?

in order to pass large structs around efficiently, without worrying about the caller mutating the value.

That doesn't apply here as you are the owner of your data.  Whereas 'in' exists to safely pass to some other piece of code elsewhere.

**Reply by ds1709** [reply]
@CyrusNajmabadi 
> 'in' exists to safely pass to some other piece of code elsewhere.

doesn't it the same for primary ctor parameters? Why safety of method parameters is important, but safety of ctor parameters is not?

**Reply by CyrusNajmabadi** [reply]
> doesn't it the same for primary ctor parameters?

You can have `in` primary ctor parameters.

**Reply by CyrusNajmabadi** [reply]
> Why safety of method parameters is important, but safety of ctor parameters is not?

you're misunderstanding the safety aspect.  If i pass my struct data to somethnig else, either they get a copy, or they get a reference.  If i want to pass by reference, i can keep that safe, by passing by 'in'.  This applies to both method and contructors.

That's a different question about how i store my own data.  I control my data and can choose how to work with it, or pass it out.

**Reply by ds1709** [reply]
@CyrusNajmabadi 
> you're misunderstanding the safety aspect.

yes, now I see.

> You can have in primary ctor parameters.

Yes, but no
```csharp
public sealed class Device(in string name, in byte address)
{
    public string Name => name; // CS9109: Cannot use ref, out, or in primary constructor parameter inside an instance member.

    public byte Address => address; // CS9109

    public int Status { get; set; }
}

public sealed class DeviceViewModel(in Device device)
{
    public string Name => device.Name; // CS9109

    public byte Address => device.Address; // CS9109

    public int Status
    {
        get => device.Status; // CS9109
        set => device.Status = value; // CS9109
    }
}
```

And a little back to `readonly` field modifier. I can make fields `readonly` to be sure it will never changed in code after class was initialized.
From Google AI:
> The purpose of a C# readonly field is to ensure that the field's value, or its reference to an object, cannot be changed after the class's constructor has completed execution. This promotes code robustness, predictability, and helps in creating immutable objects or ensuring non-nullable fields are always initialized.

I wish the same for primary ctor parameters. Especialy it looks like "MUST HAVE" feature for MVVM pattern applications. Otherwise, primary ctor looks *less safe* than regular ctor with `readonly` fields.

**Reply by CyrusNajmabadi** [reply]
> Yes, but no

Sure.  There are restrictions.  But that's very much because there would be no way to accomplish that code.  `name` is a reference to a string.  

You can trivially do this though:

```
private SomeStruct name = name; // will properly make a copy
```

That's the point.  As the caller, your data is safe.  It cannot be mutated by what you are passing into.  They can make a copy, and then mutate the copy.  But that's it.

**Reply by CyrusNajmabadi** [reply]
You are conflating mutation with capture here.  They are separate concerns.  Your data is safe in passing, regardless of if readonly is available or not for primary constructor parameters.  Receivers cannot mutate it.  They can make copies, but htat's always been the case.

--

Now, if you want to have additional protections for your *own* code, that's a reasonable thing to ask for.  But there is no safety concern.

**Reply by egvijayanand** [reply]
Suggesting an alternate approach instead of using the **in** keyword, which isn’t permitted for annotating constructor parameters.

The **ReadOnly** attribute is already defined in the BCL and can be used at any level.

https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.readonlyattribute?view=net-10.0

```cs
public class MainViewModel([ReadOnly(true)] IDialogService dialogService) : BaseViewModel
{
    // The compiler should infer this attribute and add the readonly modifier to the generated field.
    // So the field can't be changed after it's created.
}
```

**Reply by HaloFour** [reply]
> The ReadOnly attribute is already defined in the BCL and can be used at any level.

You could have an analyzer respect that attribute and warn/error if there is an attempt to reassign the parameter.

**Reply by egvijayanand** [reply]
> You could have an analyzer respect that attribute and warn/error if there is an attempt to reassign the parameter.

That kind of warning is already in place, and I’ve manually tried creating a (readonly) field to assign the PC parameter and then used it elsewhere too.

**CS9124** - Parameter is captured into the state of the enclosing type, and its value is also used to initialize a field, property, or event.

The only change is to infer the attribute, and if it’s set to true, make the compiler-generated field **readonly** so it remains immutable for the object’s lifetime.

**Reply by HaloFour** [reply]
> That kind of warning is already in place

If you're manually capturing the parameter into a field, sure.  But that's extra work on your part.

> The only change is to infer the attribute, and if it’s set to true, make the compiler-generated field readonly so it remains immutable for the object’s lifetime.

I think that's much less likely to be accepted by the language team than supporting a `readonly` modifier, and they don't seem particularly enthusiastic about that issue either.  At best, assuming they jumped on it, you wouldn't see such a feature for another year.  But you could have an analyzer that gets you the same behavior (sans the `readonly` field) within the hour.

**Comment by baterja** [comment]
I would like to be able to affect the generated field name, i.e. to allow underscore prefixes like requested here: https://github.com/dotnet/csharplang/discussions/7565

I know that primary constructor parameters are technically parameters and _might_ not be kept as fields/properties but they are in the current implementation, I even found it mentioned on the MS page: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/instance-constructors#primary-constructors

> When a primary constructor parameter is used, the compiler captures the constructor parameter in a private field with a compiler-generated name. (...) If the type includes the record modifier, the compiler instead synthesizes a public property with the same name as the primary constructor parameter.

I also find it confusing that primary constructors work differently for records than for (typical) classes, i.e. I cannot apply `field:` or `property:`-targeted attributes on PC parameters in a `class` but I can when it's a `record [class]`.

**Reply by CyrusNajmabadi** [reply]
@baterja the exact purpose of using records is precisely to get members and proper behavior around those Members.  When using a non-record type we don't want that since that's what records are for in the first place.

**Reply by baterja** [reply]
I know, but only because I've been using C# for over 10 years and was up to date when those changes were introduced. Newcomers think primary constructors will behave alike in records and not-records, and also because `TypeScript` allows similar syntax (`class X { constructor(public readonly x: number) {} }`). So I consider it counter-intuitive.

**Reply by HaloFour** [reply]
TypeScript primary constructor syntax involves providing an accessibility modifier.

The behavior of primary constructors in C# is identical to that in Scala and Kotlin, and it's not considered confusing in either of those languages.  It's expected that if you want the class to behave as a data carrier and define the schema of that data based on the constructor that you need to opt-in to that by making the class an equivalent of a record, via `case` or `data`.

**Comment by 11clock** [comment]
This feature won't be useful for me until primary constructors can have bodies for logic (like the "do" block in F#). I am mostly stuck with normal constructors for now, but if primary constructors eventually get logic bodies, I would probably be able to replace almost all of my normal constructors with primary constructors. The syntax for primary constructors is just must nicer overall.

**Comment by shawnlovelidge** [comment]
USE CASE

```
public class Tag(string name, string value, string? description = null) : Item(name, description), ITag
{
  public required string Value { get; set; } = value;
 .....
}

...

new Tag("Name", "Subscription")
      --> Error occurs "Required member [TAG.Value] must be set in the object initializer..."
```

Two issues:

(1) No body supplied for the primary constructor!!!! What genius thought this was a good idea!!!
(2) I should be able to both set a required field AND define primary constructor! This code will not compile. What's the point in having a "required" keyword if I am not allowed to satisfy the required field via the primary constructor? Morever, this annotation [SetsRequiredMembers] seems to be an afterthought.

In general, this whole attempt to shoehorn a “primary constructor” into the language has not been thought through and is incomplete at best.

I my opinion, new isn’t always better. In this case introducing a half-baked idea all for the sake of brevity is just plain stupid.

**Reply by stepanbenes** [reply]
Why do you use the `required` keyword if you are providing a value from constructor? Either remove the `required` keyword or provide value via object initializer and it will compile.

**Comment by egvijayanand** [comment]
I propose the following solution for enhancing functionality within the context of PC:

For backward compatibility, the constructor method should retain the same name as its containing type.

Historically, nearly all access modifiers have been employed to define constructors, except for the newly introduced `file`.

Thus, this new modifier can be used to define the additional method that gets invoked as part of the PC, with the invocation order determined using an attribute.

```cs
public partial class MainPage(MainViewModel viewModel) : ContentPage
{
    // Newly proposed method that acts like a constructor body, gets invoked in the PC
    file MainPage()
    {
        InitializeComponent(); // UI Initialization
        BindingContext = viewModel;
    }
}
```

**Comment by WeihanLi** [comment]
Any update for making the field `readonly`?
